packages/sdk/src/fee/FeeEstimator.ts

```diff
@@ -0,0 +1,164 @@
+
+    case "segwit":
+      return { input: 41, output: 31, txHeader: 10.5, witness: 105 };
+
+    case "p2sh-p2wpkh":
+      return { input: 64, output: 32, txHeader: 10, witness: 105 };
+
+    case "legacy":
+      return { input: 148, output: 34, txHeader: 10, witness: 0 };
+
+    default:
+      throw new OrditSDKError("Invalid type");
+  }
+}
+
+class FeeEstimator {
+  protected fee = 0;
+
+  protected feeRate: number;
+
+  protected network: Network;
+
+  protected psbt: Psbt;
+
+  protected witness?: Buffer[] = [];
+
+  protected virtualSize = 0;
+
+  protected weight = 0;
+
+  constructor({ feeRate, network, psbt, witness }: FeeEstimatorOptions) {
+    if (feeRate < 0 || !Number.isSafeInteger(feeRate)) {
+      throw new OrditSDKError("Invalid feeRate");
+    }
+
+    this.feeRate = feeRate;
+    this.network = network;
+    this.witness = witness || [];
+    this.psbt = psbt || new Psbt({ network: getNetwork(this.network) });
+  }
+
+  get data() {
+    return {
+      fee: this.fee,
+      virtualSize: this.virtualSize,
+      weight: this.weight,
+    };
+  }
+
+  calculateNetworkFee(): number {
+    this.fee = this.calculateVirtualSize() * this.feeRate;
+    // Prevents catastrophic calculations from happening
+    if (this.fee > MAXIMUM_FEE) {
+      throw new OrditSDKError("Error while calculating fees");
+    }
+
+    return this.fee;
+  }
+
+  private getInputAndOutputScriptTypes() {
+    const { inputs } = this.psbt.data;
+    const outputs = this.psbt.txOutputs;
+
+    if (inputs.length === 0) {
+      throw new OrditSDKError("PSBT must have at least one input");
+    }
+
+    if (outputs.length === 0) {
+      throw new OrditSDKError("PSBT must have at least one output");
+    }
+
+    return {
+      inputTypes: inputs.map((input) => {
+        // P2PKH (Legacy) is not supported as there is no witness script.
+        const script =
+          input.witnessUtxo && input.witnessUtxo.script
+            ? input.witnessUtxo.script
+            : null;
+
+        if (!script) {
+          throw new OrditSDKError("Invalid script");
+        }
+
+        return getScriptType(script, this.network).format;
+      }),
+      outputTypes: outputs.map(
+        (output) => getScriptType(output.script, this.network).format,
+      ),
+    };
+  }
+
+  private calculateScriptWitnessSize() {
+    const { inputTypes } = this.getInputAndOutputScriptTypes();
+    if (inputTypes.includes("taproot") && this.witness?.length) {
+      return this.witness.reduce((acc, witness) => acc + witness.byteLength, 0);
+    }
+    return 0;
+  }
+
+  private getBaseSize() {
+    const { inputTypes, outputTypes } = this.getInputAndOutputScriptTypes();
+    const witnessHeaderSize = 2;
+    const inputVBytes = inputTypes.reduce(
+      (acc, inputType) => {
+        const { input, txHeader, witness } = getBaseSizeByType(inputType);
+        acc.txHeader = txHeader;
```

Comment:
@kevzzsk raised a great point [here](https://github.com/sadoprotocol/ordit-sdk/pull/62#discussion_r1336646298), and I believe we should standardize the header size to (10.5) for all script types. At the moment, the difference is just 0.5B which is nominal and after update would only over-estimate. In many cases, the final result will be the same.

Note: Over-estimating is fine because we always check the total calculated amount against the wallet balance.

Reply:
Alright, will change this to 10.5 and make this a constant!

Reply:
Added to PR description